nomad/server.go

```diff
@@ -1197,69 +1193,44 @@ func (s *Server) setupRPC(tlsWrap tlsutil.RegionWrapper) error {
-	// that get used when component on the server is making an internal RPC
-	// call, so we only need them to be initialized once and they have no RPC
-	// context.
-	if s.staticEndpoints.Status == nil {
-		// note: Alloc, Plan have only dynamic endpoints
-		s.staticEndpoints.ACL = NewACLEndpoint(s, nil)
-		s.staticEndpoints.CSIVolume = NewCSIVolumeEndpoint(s, nil)
-		s.staticEndpoints.CSIPlugin = NewCSIPluginEndpoint(s, nil)
-		s.staticEndpoints.Deployment = NewDeploymentEndpoint(s, nil)
-		s.staticEndpoints.Job = NewJobEndpoints(s, nil)
-		s.staticEndpoints.Keyring = NewKeyringEndpoint(s, nil, s.encrypter)
-		s.staticEndpoints.Namespace = NewNamespaceEndpoint(s, nil)
-		s.staticEndpoints.Node = NewNodeEndpoint(s, nil)
-		s.staticEndpoints.Operator = NewOperatorEndpoint(s, nil)
-		s.staticEndpoints.Operator.register() // register the streaming RPCs
-		s.staticEndpoints.Periodic = NewPeriodicEndpoint(s, nil)
-		s.staticEndpoints.Region = NewRegionEndpoint(s, nil)
-		s.staticEndpoints.Scaling = NewScalingEndpoint(s, nil)
-		s.staticEndpoints.Search = NewSearchEndpoint(s, nil)
-		s.staticEndpoints.ServiceRegistration = NewServiceRegistrationEndpoint(s, nil)
-		s.staticEndpoints.Status = NewStatusEndpoint(s, nil)
-		s.staticEndpoints.System = NewSystemEndpoint(s, nil)
-		s.staticEndpoints.Variables = NewVariablesEndpoint(s, nil, s.encrypter)
-
-		s.staticEndpoints.Enterprise = NewEnterpriseEndpoints(s, nil)
-
-		// These endpoints don't have a dynamic counterpart, so they'll need to
-		// be re-registered per connection as well (see below)
-
-		// Client endpoints
-		s.staticEndpoints.ClientStats = NewClientStatsEndpoint(s)
-		s.staticEndpoints.ClientAllocations = NewClientAllocationsEndpoint(s)
-		s.staticEndpoints.ClientAllocations.register() // register the streaming RPCs
-		s.staticEndpoints.ClientCSI = NewClientCSIEndpoint(s)
-
-		// Streaming endpoints
-		s.staticEndpoints.FileSystem = NewFileSystemEndpoint(s)
-		s.staticEndpoints.FileSystem.register()
-
-		s.staticEndpoints.Agent = NewAgentEndpoint(s)
-		s.staticEndpoints.Agent.register()
-
-		s.staticEndpoints.Event = NewEventEndpoint(s)
-		s.staticEndpoints.Event.register()
-	}
-
-	// If an endpoint has any non-streaming RPCs doesn't have an RPC context,
-	// we'll register the static handler here instead of creating a new dynamic
-	// endpoint on each connection.
-
-	server.Register(s.staticEndpoints.ClientStats)
-	server.Register(s.staticEndpoints.ClientAllocations)
-	server.Register(s.staticEndpoints.ClientCSI)
-	server.Register(s.staticEndpoints.FileSystem)
-	server.Register(s.staticEndpoints.Agent)
-
-	// Dynamic endpoints are endpoints that include the connection context and
-	// are created on each connection. Register all the dynamic endpoints with
-	// the RPC server.
+	// The endpoints are client RPCs and don't include a connection
+	// context. They also need to be registered as streaming endpoints in their
+	// register() methods.
```

Comment:
I strongly considered having the constructors register the streaming RPCs here because it would make this whole function straightforward, but having that as a side effect in the constructor seems terrible. And it would make it hard to use the constructors in tests (which we've done a few places).